File: src/main/java/com/merge/api/resources/accounting/types/Item.java
package com.merge.api.resources.accounting.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = Item.Builder.class)
public final class Item {
    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<String> name;

    private final Optional<ItemStatus> status;

    private final Optional<Double> unitPrice;

    private final Optional<Double> purchasePrice;

    private final Optional<ItemPurchaseAccount> purchaseAccount;

    private final Optional<ItemSalesAccount> salesAccount;

    private final Optional<ItemCompany> company;

    private final Optional<OffsetDateTime> remoteUpdatedAt;

    private final Optional<Boolean> remoteWasDeleted;

    private final Optional<OffsetDateTime> modifiedAt;

    private final Optional<Map<String, JsonNode>> fieldMappings;

    private final Optional<List<RemoteData>> remoteData;

    private Item(
            Optional<String> id,
            Optional<String> remoteId,
            Optional<String> name,
            Optional<ItemStatus> status,
            Optional<Double> unitPrice,
            Optional<Double> purchasePrice,
            Optional<ItemPurchaseAccount> purchaseAccount,
            Optional<ItemSalesAccount> salesAccount,
            Optional<ItemCompany> company,
            Optional<OffsetDateTime> remoteUpdatedAt,
            Optional<Boolean> remoteWasDeleted,
            Optional<OffsetDateTime> modifiedAt,
            Optional<Map<String, JsonNode>> fieldMappings,
            Optional<List<RemoteData>> remoteData) {
        this.id = id;
        this.remoteId = remoteId;
        this.name = name;
        this.status = status;
        this.unitPrice = unitPrice;
        this.purchasePrice = purchasePrice;
        this.purchaseAccount = purchaseAccount;
        this.salesAccount = salesAccount;
        this.company = company;
        this.remoteUpdatedAt = remoteUpdatedAt;
        this.remoteWasDeleted = remoteWasDeleted;
        this.modifiedAt = modifiedAt;
        this.fieldMappings = fieldMappings;
        this.remoteData = remoteData;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    /**
     * @return The item's name.
     */
    @JsonProperty("name")
    public Optional<String> getName() {
        return name;
    }

    /**
     * @return The item's status.
     * <ul>
     * <li><code>ACTIVE</code> - ACTIVE</li>
     * <li><code>ARCHIVED</code> - ARCHIVED</li>
     * </ul>
     */
    @JsonProperty("status")
    public Optional<ItemStatus> getStatus() {
        return status;
    }

    /**
     * @return The item's unit price.
     */
    @JsonProperty("unit_price")
    public Optional<Double> getUnitPrice() {
        return unitPrice;
    }

    /**
     * @return The price at which the item is purchased from a vendor.
     */
    @JsonProperty("purchase_price")
    public Optional<Double> getPurchasePrice() {
        return purchasePrice;
    }

    /**
     * @return References the default account used to record a purchase of the item.
     */
    @JsonProperty("purchase_account")
    public Optional<ItemPurchaseAccount> getPurchaseAccount() {
        return purchaseAccount;
    }

    /**
     * @return References the default account used to record a sale.
     */
    @JsonProperty("sales_account")
    public Optional<ItemSalesAccount> getSalesAccount() {
        return salesAccount;
    }

    /**
     * @return The company the item belongs to.
     */
    @JsonProperty("company")
    public Optional<ItemCompany> getCompany() {
        return company;
    }

    /**
     * @return When the third party's item note was updated.
     */
    @JsonProperty("remote_updated_at")
    public Optional<OffsetDateTime> getRemoteUpdatedAt() {
        return remoteUpdatedAt;
    }

    /**
     * @return Indicates whether or not this object has been deleted by third party webhooks.
     */
    @JsonProperty("remote_was_deleted")
    public Optional<Boolean> getRemoteWasDeleted() {
        return remoteWasDeleted;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @JsonProperty("field_mappings")
    public Optional<Map<String, JsonNode>> getFieldMappings() {
        return fieldMappings;
    }

    @JsonProperty("remote_data")
    public Optional<List<RemoteData>> getRemoteData() {
        return remoteData;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof Item && equalTo((Item) other);
    }

    private boolean equalTo(Item other) {
        return id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && name.equals(other.name)
                && status.equals(other.status)
                && unitPrice.equals(other.unitPrice)
                && purchasePrice.equals(other.purchasePrice)
                && purchaseAccount.equals(other.purchaseAccount)
                && salesAccount.equals(other.salesAccount)
                && company.equals(other.company)
                && remoteUpdatedAt.equals(other.remoteUpdatedAt)
                && remoteWasDeleted.equals(other.remoteWasDeleted)
                && modifiedAt.equals(other.modifiedAt)
                && fieldMappings.equals(other.fieldMappings)
                && remoteData.equals(other.remoteData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.id,
                this.remoteId,
                this.name,
                this.status,
                this.unitPrice,
                this.purchasePrice,
                this.purchaseAccount,
                this.salesAccount,
                this.company,
                this.remoteUpdatedAt,
                this.remoteWasDeleted,
                this.modifiedAt,
                this.fieldMappings,
                this.remoteData);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<String> name = Optional.empty();

        private Optional<ItemStatus> status = Optional.empty();

        private Optional<Double> unitPrice = Optional.empty();

        private Optional<Double> purchasePrice = Optional.empty();

        private Optional<ItemPurchaseAccount> purchaseAccount = Optional.empty();

        private Optional<ItemSalesAccount> salesAccount = Optional.empty();

        private Optional<ItemCompany> company = Optional.empty();

        private Optional<OffsetDateTime> remoteUpdatedAt = Optional.empty();

        private Optional<Boolean> remoteWasDeleted = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Optional<Map<String, JsonNode>> fieldMappings = Optional.empty();

        private Optional<List<RemoteData>> remoteData = Optional.empty();

        private Builder() {}

        public Builder from(Item other) {
            id(other.getId());
            remoteId(other.getRemoteId());
            name(other.getName());
            status(other.getStatus());
            unitPrice(other.getUnitPrice());
            purchasePrice(other.getPurchasePrice());
            purchaseAccount(other.getPurchaseAccount());
            salesAccount(other.getSalesAccount());
            company(other.getCompany());
            remoteUpdatedAt(other.getRemoteUpdatedAt());
            remoteWasDeleted(other.getRemoteWasDeleted());
            modifiedAt(other.getModifiedAt());
            fieldMappings(other.getFieldMappings());
            remoteData(other.getRemoteData());
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "name", nulls = Nulls.SKIP)
        public Builder name(Optional<String> name) {
            this.name = name;
            return this;
        }

        public Builder name(String name) {
            this.name = Optional.of(name);
            return this;
        }

        @JsonSetter(value = "status", nulls = Nulls.SKIP)
        public Builder status(Optional<ItemStatus> status) {
            this.status = status;
            return this;
        }

        public Builder status(ItemStatus status) {
            this.status = Optional.of(status);
            return this;
        }

        @JsonSetter(value = "unit_price", nulls = Nulls.SKIP)
        public Builder unitPrice(Optional<Double> unitPrice) {
            this.unitPrice = unitPrice;
            return this;
        }

        public Builder unitPrice(Double unitPrice) {
            this.unitPrice = Optional.of(unitPrice);
            return this;
        }

        @JsonSetter(value = "purchase_price", nulls = Nulls.SKIP)
        public Builder purchasePrice(Optional<Double> purchasePrice) {
            this.purchasePrice = purchasePrice;
            return this;
        }

        public Builder purchasePrice(Double purchasePrice) {
            this.purchasePrice = Optional.of(purchasePrice);
            return this;
        }

        @JsonSetter(value = "purchase_account", nulls = Nulls.SKIP)
        public Builder purchaseAccount(Optional<ItemPurchaseAccount> purchaseAccount) {
            this.purchaseAccount = purchaseAccount;
            return this;
        }

        public Builder purchaseAccount(ItemPurchaseAccount purchaseAccount) {
            this.purchaseAccount = Optional.of(purchaseAccount);
            return this;
        }

        @JsonSetter(value = "sales_account", nulls = Nulls.SKIP)
        public Builder salesAccount(Optional<ItemSalesAccount> salesAccount) {
            this.salesAccount = salesAccount;
            return this;
        }

        public Builder salesAccount(ItemSalesAccount salesAccount) {
            this.salesAccount = Optional.of(salesAccount);
            return this;
        }

        @JsonSetter(value = "company", nulls = Nulls.SKIP)
        public Builder company(Optional<ItemCompany> company) {
            this.company = company;
            return this;
        }

        public Builder company(ItemCompany company) {
            this.company = Optional.of(company);
            return this;
        }

        @JsonSetter(value = "remote_updated_at", nulls = Nulls.SKIP)
        public Builder remoteUpdatedAt(Optional<OffsetDateTime> remoteUpdatedAt) {
            this.remoteUpdatedAt = remoteUpdatedAt;
            return this;
        }

        public Builder remoteUpdatedAt(OffsetDateTime remoteUpdatedAt) {
            this.remoteUpdatedAt = Optional.of(remoteUpdatedAt);
            return this;
        }

        @JsonSetter(value = "remote_was_deleted", nulls = Nulls.SKIP)
        public Builder remoteWasDeleted(Optional<Boolean> remoteWasDeleted) {
            this.remoteWasDeleted = remoteWasDeleted;
            return this;
        }

        public Builder remoteWasDeleted(Boolean remoteWasDeleted) {
            this.remoteWasDeleted = Optional.of(remoteWasDeleted);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        @JsonSetter(value = "field_mappings", nulls = Nulls.SKIP)
        public Builder fieldMappings(Optional<Map<String, JsonNode>> fieldMappings) {
            this.fieldMappings = fieldMappings;
            return this;
        }

        public Builder fieldMappings(Map<String, JsonNode> fieldMappings) {
            this.fieldMappings = Optional.of(fieldMappings);
            return this;
        }

        @JsonSetter(value = "remote_data", nulls = Nulls.SKIP)
        public Builder remoteData(Optional<List<RemoteData>> remoteData) {
            this.remoteData = remoteData;
            return this;
        }

        public Builder remoteData(List<RemoteData> remoteData) {
            this.remoteData = Optional.of(remoteData);
            return this;
        }

        public Item build() {
            return new Item(
                    id,
                    remoteId,
                    name,
                    status,
                    unitPrice,
                    purchasePrice,
                    purchaseAccount,
                    salesAccount,
                    company,
                    remoteUpdatedAt,
                    remoteWasDeleted,
                    modifiedAt,
                    fieldMappings,
                    remoteData);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/accounting/Item.kt
package dev.merge.api.models.accounting

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.errors.MergeInvalidDataException
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The Item Object
 *
 * ### Description
 *
 * The `Item` object refers to the goods involved in a transaction.
 *
 * ### Usage Example
 *
 * Fetch from the `LIST Items` endpoint and view a company's items.
 */
@JsonDeserialize(builder = Item.Builder::class)
@NoAutoDetect
class Item
private constructor(
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val name: JsonField<String>,
    private val status: JsonField<Status7d1Enum>,
    private val unitPrice: JsonField<Double>,
    private val purchasePrice: JsonField<Double>,
    private val purchaseAccount: JsonField<String>,
    private val salesAccount: JsonField<String>,
    private val company: JsonField<String>,
    private val remoteUpdatedAt: JsonField<OffsetDateTime>,
    private val remoteWasDeleted: JsonField<Boolean>,
    private val fieldMappings: JsonValue,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val remoteData: JsonField<List<RemoteData>>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /** The item's name. */
    fun name(): Optional<String> = Optional.ofNullable(name.getNullable("name"))

    /**
     * - `ACTIVE` - ACTIVE
     * - `ARCHIVED` - ARCHIVED
     */
    fun status(): Optional<Status7d1Enum> = Optional.ofNullable(status.getNullable("status"))

    /** The item's unit price. */
    fun unitPrice(): Optional<Double> = Optional.ofNullable(unitPrice.getNullable("unit_price"))

    /** The price at which the item is purchased from a vendor. */
    fun purchasePrice(): Optional<Double> =
        Optional.ofNullable(purchasePrice.getNullable("purchase_price"))

    /** References the default account used to record a purchase of the item. */
    fun purchaseAccount(): Optional<String> =
        Optional.ofNullable(purchaseAccount.getNullable("purchase_account"))

    /** References the default account used to record a sale. */
    fun salesAccount(): Optional<String> =
        Optional.ofNullable(salesAccount.getNullable("sales_account"))

    /** The company the item belongs to. */
    fun company(): Optional<String> = Optional.ofNullable(company.getNullable("company"))

    /** When the third party's item note was updated. */
    fun remoteUpdatedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(remoteUpdatedAt.getNullable("remote_updated_at"))

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    fun remoteWasDeleted(): Optional<Boolean> =
        Optional.ofNullable(remoteWasDeleted.getNullable("remote_was_deleted"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    fun remoteData(): Optional<List<RemoteData>> =
        Optional.ofNullable(remoteData.getNullable("remote_data"))

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    /** The item's name. */
    @JsonProperty("name") @ExcludeMissing fun _name() = name

    /**
     * - `ACTIVE` - ACTIVE
     * - `ARCHIVED` - ARCHIVED
     */
    @JsonProperty("status") @ExcludeMissing fun _status() = status

    /** The item's unit price. */
    @JsonProperty("unit_price") @ExcludeMissing fun _unitPrice() = unitPrice

    /** The price at which the item is purchased from a vendor. */
    @JsonProperty("purchase_price") @ExcludeMissing fun _purchasePrice() = purchasePrice

    /** References the default account used to record a purchase of the item. */
    @JsonProperty("purchase_account") @ExcludeMissing fun _purchaseAccount() = purchaseAccount

    /** References the default account used to record a sale. */
    @JsonProperty("sales_account") @ExcludeMissing fun _salesAccount() = salesAccount

    /** The company the item belongs to. */
    @JsonProperty("company") @ExcludeMissing fun _company() = company

    /** When the third party's item note was updated. */
    @JsonProperty("remote_updated_at") @ExcludeMissing fun _remoteUpdatedAt() = remoteUpdatedAt

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    @JsonProperty("remote_was_deleted") @ExcludeMissing fun _remoteWasDeleted() = remoteWasDeleted

    @JsonProperty("field_mappings") @ExcludeMissing fun _fieldMappings() = fieldMappings

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonProperty("remote_data") @ExcludeMissing fun _remoteData() = remoteData

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate() = apply {
        if (!validated) {
            id()
            remoteId()
            name()
            status()
            unitPrice()
            purchasePrice()
            purchaseAccount()
            salesAccount()
            company()
            remoteUpdatedAt()
            remoteWasDeleted()
            modifiedAt()
            remoteData().map { it.forEach { it.validate() } }
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is Item &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.name == other.name &&
            this.status == other.status &&
            this.unitPrice == other.unitPrice &&
            this.purchasePrice == other.purchasePrice &&
            this.purchaseAccount == other.purchaseAccount &&
            this.salesAccount == other.salesAccount &&
            this.company == other.company &&
            this.remoteUpdatedAt == other.remoteUpdatedAt &&
            this.remoteWasDeleted == other.remoteWasDeleted &&
            this.fieldMappings == other.fieldMappings &&
            this.modifiedAt == other.modifiedAt &&
            this.remoteData == other.remoteData &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    id,
                    remoteId,
                    name,
                    status,
                    unitPrice,
                    purchasePrice,
                    purchaseAccount,
                    salesAccount,
                    company,
                    remoteUpdatedAt,
                    remoteWasDeleted,
                    fieldMappings,
                    modifiedAt,
                    remoteData,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "Item{id=$id, remoteId=$remoteId, name=$name, status=$status, unitPrice=$unitPrice, purchasePrice=$purchasePrice, purchaseAccount=$purchaseAccount, salesAccount=$salesAccount, company=$company, remoteUpdatedAt=$remoteUpdatedAt, remoteWasDeleted=$remoteWasDeleted, fieldMappings=$fieldMappings, modifiedAt=$modifiedAt, remoteData=$remoteData, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var name: JsonField<String> = JsonMissing.of()
        private var status: JsonField<Status7d1Enum> = JsonMissing.of()
        private var unitPrice: JsonField<Double> = JsonMissing.of()
        private var purchasePrice: JsonField<Double> = JsonMissing.of()
        private var purchaseAccount: JsonField<String> = JsonMissing.of()
        private var salesAccount: JsonField<String> = JsonMissing.of()
        private var company: JsonField<String> = JsonMissing.of()
        private var remoteUpdatedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteWasDeleted: JsonField<Boolean> = JsonMissing.of()
        private var fieldMappings: JsonValue = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteData: JsonField<List<RemoteData>> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(item: Item) = apply {
            this.id = item.id
            this.remoteId = item.remoteId
            this.name = item.name
            this.status = item.status
            this.unitPrice = item.unitPrice
            this.purchasePrice = item.purchasePrice
            this.purchaseAccount = item.purchaseAccount
            this.salesAccount = item.salesAccount
            this.company = item.company
            this.remoteUpdatedAt = item.remoteUpdatedAt
            this.remoteWasDeleted = item.remoteWasDeleted
            this.fieldMappings = item.fieldMappings
            this.modifiedAt = item.modifiedAt
            this.remoteData = item.remoteData
            additionalProperties(item.additionalProperties)
        }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        /** The item's name. */
        fun name(name: String) = name(JsonField.of(name))

        /** The item's name. */
        @JsonProperty("name")
        @ExcludeMissing
        fun name(name: JsonField<String>) = apply { this.name = name }

        /**
         * - `ACTIVE` - ACTIVE
         * - `ARCHIVED` - ARCHIVED
         */
        fun status(status: Status7d1Enum) = status(JsonField.of(status))

        /**
         * - `ACTIVE` - ACTIVE
         * - `ARCHIVED` - ARCHIVED
         */
        @JsonProperty("status")
        @ExcludeMissing
        fun status(status: JsonField<Status7d1Enum>) = apply { this.status = status }

        /** The item's unit price. */
        fun unitPrice(unitPrice: Double) = unitPrice(JsonField.of(unitPrice))

        /** The item's unit price. */
        @JsonProperty("unit_price")
        @ExcludeMissing
        fun unitPrice(unitPrice: JsonField<Double>) = apply { this.unitPrice = unitPrice }

        /** The price at which the item is purchased from a vendor. */
        fun purchasePrice(purchasePrice: Double) = purchasePrice(JsonField.of(purchasePrice))

        /** The price at which the item is purchased from a vendor. */
        @JsonProperty("purchase_price")
        @ExcludeMissing
        fun purchasePrice(purchasePrice: JsonField<Double>) = apply {
            this.purchasePrice = purchasePrice
        }

        /** References the default account used to record a purchase of the item. */
        fun purchaseAccount(purchaseAccount: String) =
            purchaseAccount(JsonField.of(purchaseAccount))

        /** References the default account used to record a purchase of the item. */
        @JsonProperty("purchase_account")
        @ExcludeMissing
        fun purchaseAccount(purchaseAccount: JsonField<String>) = apply {
            this.purchaseAccount = purchaseAccount
        }

        /** References the default account used to record a sale. */
        fun salesAccount(salesAccount: String) = salesAccount(JsonField.of(salesAccount))

        /** References the default account used to record a sale. */
        @JsonProperty("sales_account")
        @ExcludeMissing
        fun salesAccount(salesAccount: JsonField<String>) = apply {
            this.salesAccount = salesAccount
        }

        /** The company the item belongs to. */
        fun company(company: String) = company(JsonField.of(company))

        /** The company the item belongs to. */
        @JsonProperty("company")
        @ExcludeMissing
        fun company(company: JsonField<String>) = apply { this.company = company }

        /** When the third party's item note was updated. */
        fun remoteUpdatedAt(remoteUpdatedAt: OffsetDateTime) =
            remoteUpdatedAt(JsonField.of(remoteUpdatedAt))

        /** When the third party's item note was updated. */
        @JsonProperty("remote_updated_at")
        @ExcludeMissing
        fun remoteUpdatedAt(remoteUpdatedAt: JsonField<OffsetDateTime>) = apply {
            this.remoteUpdatedAt = remoteUpdatedAt
        }

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        fun remoteWasDeleted(remoteWasDeleted: Boolean) =
            remoteWasDeleted(JsonField.of(remoteWasDeleted))

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        @JsonProperty("remote_was_deleted")
        @ExcludeMissing
        fun remoteWasDeleted(remoteWasDeleted: JsonField<Boolean>) = apply {
            this.remoteWasDeleted = remoteWasDeleted
        }

        @JsonProperty("field_mappings")
        @ExcludeMissing
        fun fieldMappings(fieldMappings: JsonValue) = apply { this.fieldMappings = fieldMappings }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun remoteData(remoteData: List<RemoteData>) = remoteData(JsonField.of(remoteData))

        @JsonProperty("remote_data")
        @ExcludeMissing
        fun remoteData(remoteData: JsonField<List<RemoteData>>) = apply {
            this.remoteData = remoteData
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): Item =
            Item(
                id,
                remoteId,
                name,
                status,
                unitPrice,
                purchasePrice,
                purchaseAccount,
                salesAccount,
                company,
                remoteUpdatedAt,
                remoteWasDeleted,
                fieldMappings,
                modifiedAt,
                remoteData.map { it.toUnmodifiable() },
                additionalProperties.toUnmodifiable(),
            )
    }

    class Status7d1Enum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is Status7d1Enum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val ACTIVE = Status7d1Enum(JsonField.of("ACTIVE"))

            @JvmField val ARCHIVED = Status7d1Enum(JsonField.of("ARCHIVED"))

            @JvmStatic fun of(value: String) = Status7d1Enum(JsonField.of(value))
        }

        enum class Known {
            ACTIVE,
            ARCHIVED,
        }

        enum class Value {
            ACTIVE,
            ARCHIVED,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                ACTIVE -> Value.ACTIVE
                ARCHIVED -> Value.ARCHIVED
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                ACTIVE -> Known.ACTIVE
                ARCHIVED -> Known.ARCHIVED
                else -> throw MergeInvalidDataException("Unknown Status7d1Enum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }
}
